### PR TITLE
Handle plugins under the test scope

### PR DESCRIPTION
**Pull request by kwhetstone**
Some of the plugins are failing on the newer versions of Jenkins because their test elements need to be updated.  These were failing pretty much immediately, so I changed the scope to also cover the tests.

Also found that there are instances where the bundled elements are supposed to be upgraded but aren't actually upgraded because they weren't included in the base pom.  These are now explicitly added.

@reviewbybees

### CODE REVIEW

**Comment by kwhetstone** [comment]
@reviewbybees done

I'd appreciate if someone who's familiar with the pct to take a look.  I've run these through several tests with different plugins, especially with the new war files.  This will help us get better results from our tests.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 1]
```diff
@@ -539,39 +539,67 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
...
+                    String groupId;
+                    String artifactId;
                     VersionNumber version;
-                    try {
-                        version = new VersionNumber(m.group(4));
-                    } catch (IllegalArgumentException x) {
-                        // OK, some other kind of dep, just ignore
+                    if (!m.matches() && !m2.matches()) {
                         continue;
+                    } else if (m.matches()) {
```
Comment (varyvol) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Maybe this logic could be extracted into a method, as it's the same in both if clauses.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 2]
```diff
@@ -639,35 +669,43 @@ private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mc
                     toAdd.put(plugin, declaredMinimum);
                 }
             }
-            for (Map.Entry<String,VersionNumber> pluginDep : pluginDeps.entrySet()) {
+
+            checkDefinedDeps(pluginDeps, toAdd, toReplace, otherPlugins);
+            pluginDepsTest.putAll(toAdd);
+            pluginDepsTest.putAll(toReplace);
+            checkDefinedDeps(pluginDepsTest, toAddTest, toReplaceTest, otherPlugins);
+            if (!toAdd.isEmpty() || !toReplace.isEmpty()) {
```
Comment (varyvol) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Don't you need to check here the test ones?

**Comment by varyvol** [comment]
@kwhetstone I made a couple of minor (I think) comments.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 1 above)
Comment (kwhetstone) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: This is definitely a situation where I wish I could have used Groovy.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java (hunk 2 above)
Comment (kwhetstone) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Yes, good catch.  Added.

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 3]
```diff
@@ -341,7 +341,7 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
 			ScmManager scmManager = SCMManagerFactory.getInstance().createScmManager();
 			ScmRepository repository = scmManager.makeScmRepository(pomData.getConnectionUrl());
 			CheckOutScmResult result = scmManager.checkOut(repository, new ScmFileSet(pluginCheckoutDir), new ScmTag(plugin.name+"-"+plugin.version));
-			
+			System.out.println("DONE?");
```
Comment (andresrc) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Are we leaving these messages in?

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 4]
```diff
@@ -390,7 +391,7 @@ private TestExecutionResult testPluginAgainst(MavenCoordinates coreCoordinates,
             // This defends against source incompatibilities (which we do not care about for this purpose);
             // and ensures that we are testing a plugin binary as close as possible to what was actually released.
             // We also skip potential javadoc execution to avoid general test failure.
-            runner.run(mconfig, pluginCheckoutDir, buildLogFile, "clean", "process-test-classes", "-Dmaven.javadoc.skip");
+            //runner.run(mconfig, pluginCheckoutDir, buildLogFile, "clean", "process-test-classes", "-Dmaven.javadoc.skip");
```
Comment (andresrc) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Where is it run now?

File: plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java [hunk 5]
```diff
@@ -639,35 +667,74 @@ private void addSplitPluginDependencies(String thisPlugin, MavenRunner.Config mc
...
+    /**
+     * Search the dependents of a given plugin to determine if they need to be converted.
+     */
+    private void updateAllDependents(String parent, Plugin dependent, Map<String,VersionNumber> pluginList, Map<String,VersionNumber> adding, Map<String,VersionNumber> replacing, Map<String,Plugin> otherPlugins, List<String> inTest, List<String> toConvertFromTest) {
+        // Check if this exists with an undesired scope
+        String pluginName = dependent.name;
+        if(inTest.contains(pluginName)) { // becomes a replacement 
+            System.out.println("Converting " + pluginName + " from the test scope since it was a dependency of " + parent);
+            toConvertFromTest.add(pluginName);
+            replacing.put(pluginName, new VersionNumber(dependent.version));
```
Comment (andresrc) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Can you point to a plugin having this issue to better understand the case?

Comment (kwhetstone) [inline plugins-compat-tester/src/main/java/org/jenkins/tools/test/PluginCompatTester.java]: Sure.  The first one I can think of off the top of my head is the copyartifact plugin.  I can add a reference to it in a comment.

**Comment by kwhetstone** [comment]
I added some clarification.  Essentially, this new change recursively searches the upgraded plugins to also update their dependencies.  This more accurately reflects the state of a Jenkins war install; when a plugin is installed it will pull the latest versions of all dependencies with it.  This also updates all the included test dependencies since those are also important to keeping the plugin-under-test relatively free of errors.

**Pull request merged**